File: TypeScript/lexer.ts
import { readFileSync } from "fs";

export enum Role {
    Identifier,
    Equals,
    OpenParent,
    CloseParent,
    OpenBrace,
    CloseBrace,
    OpenBrack,
    CloseBrack,
    Number,
    PlusOperator,
    MinusOperator,
    MultOperator,
    DivOperator,
    EndLine,
    TypingInteger,
    TypingBoolean,
    TypingString,
    TypingFloat,
    FunctionDeclaration,
    ForLoopDeclaration,
    WhileLoopDeclaration,
    DoLoopDeclaration,
    String,
    ColonOperator,
    ModOperator,
    Null,
    Comma,
    BoolT,
    BoolF,
    LetDeclare,
    ConstDeclare,
    EndFile,
    IfCond,
    ElsifCond,
    ElseCond,
    SwitchCond,
    CaseCond,
}

export const KeyWords: Record<string, Role> = {
    "Ghen": Role.PlusOperator,
    "gHen": Role.MinusOperator,
    "ghEn": Role.MultOperator,
    "gheN": Role.DivOperator,
    "GHEN": Role.ModOperator,
    "joshua": Role.FunctionDeclaration,
    "bax": Role.WhileLoopDeclaration,
    "ter": Role.DoLoopDeclaration,
    "baxter": Role.ForLoopDeclaration,
    "carloss": Role.Null,
    "ghen": Role.Equals,
    "Zer": Role.BoolT,
    "Xeus": Role.BoolF,
    "car": Role.LetDeclare,
    "LOS": Role.ConstDeclare,
    "BAX": Role.IfCond,
    "BAXTER" : Role.ElsifCond,
    "TER" : Role.ElseCond,
    "BAXSTAR" : Role.SwitchCond,
    "star" : Role.CaseCond
}

function alphaCheck(value: string): boolean  {
    return value.toUpperCase() != value.toLowerCase();
}

export interface Word {
    value: string,
    role: Role
}

export function makeWord(value = "", role: Role): Word { return { value, role }; }

export function analyze(codes: string): Word[] {
    const words = new Array<Word>();
    const src = codes.split('');
    while (src.length > 0) {
        if(src[0] == '(') words.push(makeWord(src.shift(), Role.OpenParent));
        else if(src[0] == ')') words.push(makeWord(src.shift(), Role.CloseParent));
        else if(src[0] == '[') words.push(makeWord(src.shift(), Role.OpenBrack));
        else if(src[0] == ']') words.push(makeWord(src.shift(), Role.CloseBrack));
        else if(src[0] == '{') words.push(makeWord(src.shift(), Role.OpenBrace));
        else if(src[0] == '}') words.push(makeWord(src.shift(), Role.CloseBrace));
        else if(src[0] == '=') words.push(makeWord(src.shift(), Role.Equals));
        else if(src[0] == ',') words.push(makeWord(src.shift(), Role.Comma));
        else if(src[0] == ';') words.push(makeWord(src.shift(), Role.EndLine));
        else {
            //for multicharacter words
            if(!(isNaN(parseInt(src[0])))) {
                let numval = "";
                while(src.length > 0 && !(isNaN(parseInt(src[0])))) numval += src.shift();
                words.push(makeWord(numval, Role.Number));
            }
            else if(src[0] == '"') {
                src.shift();
                let string = "";
                while(src.length > 0 && src[0] != '"') string += src.shift();
                words.push(makeWord(string, Role.String));
                src.shift();
            } 
            else if(alphaCheck(src[0])) {
                let id = "";
                while(src.length > 0 && alphaCheck(src[0])) id += src.shift();
                words.push(makeWord(id, KeyWords[id] == undefined ? Role.Identifier : KeyWords[id]));
            }
            else if(src[0] == ' ' || src[0] == '\n' || src[0] == '\t' || src[0] == '\r') src.shift();
            else {
                throw `Unknown character/word was found: ${src}`;
                process.exit();
            }
        }
    }
    words.push(makeWord("EOF", Role.EndFile));
    return words;
}

// const code = readFileSync("./hello.cbjg", 'utf-8');
// for(const word of analyze(code)) console.log(word);

File: Lua/lexer.lua
Ore = { rock=nil, mineral=nil }
local keywords = {
    ["Ghen"] = "PlusRock",
    ["gHen"] = "MinusRock",
    ["ghEn"] = "MultRock",
    ["gheN"] = "DivRock",
    ["GHEN"] = "ModRock",
    ["ghen"] = "EqualsRock",
    ["car"] = "DeclareRock",
    ["LOS"] = "ConstRock",
}

function Ore.new(rock, mineral)
    local self = setmetatable({}, Ore)
    self.rock = rock
    self.mineral = mineral
    return self
end

function minerocks(code) 
    local ores = {}
    local oreindex = 1
    local i = 1
    while i <= #code do
        if code:sub(i, i) == "(" then ores[oreindex] = Ore.new("OpenParent", "(") oreindex = oreindex+1
        elseif code:sub(i, i) == ")" then ores[oreindex] = Ore.new("CloseParent", ")") oreindex = oreindex+1
        elseif code:sub(i, i) == "=" then ores[oreindex] = Ore.new("EqualsRock", "=") oreindex = oreindex+1
        elseif code:sub(i, i) == "," then ores[oreindex] = Ore.new("CommaRock", ",") oreindex = oreindex+1
        else
            local orestring = ""
            if code:sub(i, i) == "\"" then
                i=i+1
                while code:sub(i, i) ~= "\"" and i <= #code do orestring=orestring..code:sub(i, i) i=i+1 end
                ores[oreindex] = Ore.new("StringRock", orestring)
                oreindex = oreindex+1
            elseif code:sub(i, i):match("%d") then
                while code:sub(i, i):match("%d") and i <= #code do orestring=orestring..code:sub(i, i) i=i+1 end
                ores[oreindex] = Ore.new("NumberRock", orestring)
                oreindex = oreindex+1
                i=i-1
            elseif code:sub(i, i):match("%w") then
                while code:sub(i, i):match("%w") and keywords[orestring] == nil and i <= #code do orestring=orestring..code:sub(i, i) i=i+1 end
                if keywords[orestring] then ores[oreindex] = Ore.new(keywords[orestring], orestring)
                else ores[oreindex] = Ore.new("Identifier", orestring) end
                oreindex = oreindex+1
                i=i-1
            else code:sub(i, i):match("%s")
            end
        end
        i=i+1
    end
    ores[oreindex] = Ore.new("EndFile", "EOF")
    return ores
end